src/nvim/ex_cmds.c

```diff
@@ -4013,9 +4019,13 @@ void do_sub(exarg_T *eap)
           p1 = ml_get(sub_firstlnum + nmatch - 1);
           nmatch_tl += nmatch - 1;
         }
-        copy_len = regmatch.startpos[0].col - copycol;
-        needed_len = copy_len + ((unsigned)STRLEN(p1)
-                                 - regmatch.endpos[0].col) + sublen + 1;
+        colnr_T startcol = regmatch.startpos[0].col;
+        assert(startcol >= copycol);
+        copy_len = (size_t)(startcol - copycol);
+        colnr_T endcol = regmatch.endpos[0].col;
+        colnr_T extra_len = ((colnr_T) STRLEN(p1) - endcol) + sublen + 1;
+        assert(extra_len >= 0);
+        needed_len = copy_len + (size_t) extra_len;
```

Comment:
Again, why the need of the new variables instead of the in-place casts? I'm genuinely curious.